array-manipulation-1/src/compact.js

```diff
@@ -0,0 +1,11 @@
+/* exported compact */
+function compact(array) {
+  var newArray = [];
+
+  for (var i = 0; i < array.length; i++) {
+    if (typeof array[i] !== 'undefined' && array[i] !== false && array[i]) {
```

Comment:
`typeof array[i] !== 'undefined' && array[i] !== false && array[i]`

Can you explain your reasoning behind this condition? I believe this is the same format we discussed over Slack earlier, did you make any changes? Do you know why this works? If not, please complete this in a way that you fully understand.

Reply:
So the test is designed to test for the following data types:

1.  'True' boolean data types.
2.  'Null" data types.
3. "NaN" data types.
4. Data types that are classified as 'undefined'
5.  "empty strings" or ("")

Now, to my understanding, these, like all data types, evaluate to a "TRUTHY" or "FALSEY" evaluation when they are put to the test in a true or false evaluation. Or that is to say, a strict " === " evaluation. That means in the If statement on line 6, they are put through 3 tests.

1. FIRST - is the array item currently being iterated over 'undefined'?
2. SECOND - is the array item currently being iterated over 'false'? 
3. THIRD - Since all if statements default to testing for 'TRUE' if no test is put to them, than they default to checking for TRUE. This means that the Second statement is redundant. I just tested this to confirm it, and it was. I also removed the first statement, and it also worked. so the first and second statements were redundant. In my submission I'm going to remove them and just leave the (array[i]) evaluation in there.

Now, you asked why this works. On MDN it says, "In JavaScript a truthy value is a value that is considered true when encountered in a boolean context. All values are truthy unless they are defined as falsy. **(i.e. expect for false, 0, -0, On, "", null, undefined and NaN.**

So the reason this works is because the data types that you wanted removed are designed to be evaluated is falsey and in the 'if else'  statement, when no other test is provided, it defaults to testing for truthy/falsey.

Reply:
![image](https://user-images.githubusercontent.com/51938797/112377194-ad98e080-8ca2-11eb-9a46-7dfcd8496fab.png)

Reply:
@Steve-A-Dore Great explanation! 👏 👏 👏